File: l04.go
package main


// 作者：LeetCode-Solution
// 链接：https://leetcode-cn.com/problems/median-of-two-sorted-arrays/solution/xun-zhao-liang-ge-you-xu-shu-zu-de-zhong-wei-s-114/
// 来源：力扣（LeetCode）
// 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。


func main(){
	nums1 := []int{1,3}
	nums2 := []int{2,7}
	println("findMedianSortedArrays", findMedianSortedArrays(nums1, nums2))
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    totalLength := len(nums1) + len(nums2)
    if totalLength%2 == 1 {
        midIndex := totalLength/2
        return float64(getKthElement(nums1, nums2, midIndex + 1))
    } else {
        midIndex1, midIndex2 := totalLength/2 - 1, totalLength/2
        return float64(getKthElement(nums1, nums2, midIndex1 + 1) + getKthElement(nums1, nums2, midIndex2 + 1)) / 2.0
    }
    return 0
}

func getKthElement(nums1, nums2 []int, k int) int {
    index1, index2 := 0, 0
    for {
        if index1 == len(nums1) {
            return nums2[index2 + k - 1]
        }
        if index2 == len(nums2) {
            return nums1[index1 + k - 1]
        }
        if k == 1 {
            return min(nums1[index1], nums2[index2])
        }
        half := k/2
        newIndex1 := min(index1 + half, len(nums1)) - 1
        newIndex2 := min(index2 + half, len(nums2)) - 1
        pivot1, pivot2 := nums1[newIndex1], nums2[newIndex2]
        if pivot1 <= pivot2 {
            k -= (newIndex1 - index1 + 1)
            index1 = newIndex1 + 1
        } else {
            k -= (newIndex2 - index2 + 1)
            index2 = newIndex2 + 1
        }
    }
    return 0
}

func min(x, y int) int {
    if x < y {
        return x
    }
    return y
}

File: c/l04.c
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include <string.h>

#define MIN(a, b) (a < b ? a : b)
#define MAX(a, b) (a > b ? a : b)

// 4 https://leetcode-cn.com/problems/median-of-two-sorted-arrays
// 给定两个大小分别为 m 和 n 的正序（从小到大）数组 nums1 和 nums2。请你找出并返回这两个正序数组的 中位数 。
double getKthArrays(int* nums1, int l1, int* nums2, int l2, int k){
    int i1 = 0;
    int i2 = 0;
    int h = 0;
    int ni1 = 0;
    int ni2 = 0;
    int p1 = 0;
    int p2 = 0;
    while(1){
        if(i1 == l1) return nums2[i2+k-1];
        if(i2 == l2) return nums1[i1+k-1];
        if(k == 1)   return MIN(nums1[i1], nums2[i2]);

        h = k/2;
        ni1 = MIN(i1 + h, l1) - 1;
        ni2 = MIN(i2 + h, l2) - 1;
        p1 = nums1[ni1];
        p2 = nums2[ni2];
        if(p1 <= p2){
            k -= (ni1 - i1 + 1);
            i1 = ni1 + 1;
        }else{
            k -= (ni2 - i2 + 1);
            i2 = ni2 + 1;
        }
    }
    return 0;
}
double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2, int nums2Size){
    int s = nums1Size + nums2Size;
    if(s%2 == 1){
        return getKthArrays(nums1, nums1Size, nums2, nums2Size, s/2 + 1);
    }
    else return   getKthArrays(nums1, nums1Size, nums2, nums2Size, s/2  )/2.0f 
                + getKthArrays(nums1, nums1Size, nums2, nums2Size, s/2+1)/2.0f;
}

int main(int c, char** vv){
    int ns1[] = {1,3};
    int ns2[] = {2,7};
    int l1 = sizeof(ns1)/sizeof(int);
    int l2 = sizeof(ns2)/sizeof(int);
    printf("%f\n", findMedianSortedArrays(ns1, l1, ns2, l2));
    return 0;
}